### PR TITLE
Write correct header size when encoding task headers

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

When we write the headers size we would include invalid headers. These headers would be filtered and not written to the buffer. Therefore, the size could differ from the actual amount of headers written in the buffer.

With this fix we first validate all the headers, before writing the header size. This way we will never write the wrong size.

## Related issues

Relates to #8611

Since I have not yet been able to reproduce the problem I have decided against closing this issue with this PR. First I'd like to discuss if this could indeed be the root cause.

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [x] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Review by romansmirnov — CHANGES_REQUESTED** [review]
@remcowesterhoud,

:question: do you think, it could make sense to log a warning (or a debug log) when not all headers are written? In case of a subsequent failure (because an expected header is missing or it didn't behave correctly), it could help to trace it. WDYT?

**Comment by romansmirnov** [comment]
Another :question: Am I correct, that there will be a separate PR to recover from that state? (see https://github.com/camunda-cloud/zeebe/issues/8611#issuecomment-1020309495)

**Comment by remcowesterhoud** [comment]
@romansmirnov Thanks for your review! I think it's a valid point so I've added the logging. Now that we know how to reproduce I've also included a unit test.

As for your other question. I didn't take this as part of this PR, that's also why this PR doesn't close the issue yet. I'll have to think of a proper way to recover this.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/UserTaskTest.java [hunk 1]
```diff
@@ -299,6 +299,31 @@ public void shouldCreateJobWithEvaluatedAssigneeExpressionHeader() {
...
+    // then
+    final Record<JobRecordValue> job =
+        RecordingExporter.jobRecords(JobIntent.CREATED)
+            .withProcessInstanceKey(processInstanceKey)
+            .getFirst();
+
+    final Map<String, String> customHeaders = job.getValue().getCustomHeaders();
+    assertThat(customHeaders)
+        .hasSize(0)
+        .doesNotContainEntry(Protocol.USER_TASK_ASSIGNEE_HEADER_NAME, "alice");
```
Comment (oleschoenburg) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/UserTaskTest.java:324]: Sorry for the drive-by review but this looks like a copy-paste mistake. It should probably be something like `doesNotContainKey(Protocol.USER_TASK_ASSIGNEE_HEADER_NAME)`

Comment (remcowesterhoud) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/UserTaskTest.java:324]: You are correct 🙈 I'll modify it. It still tests the same thing because of the `hasSize(0)` but this is definitely more clear.

**Review by romansmirnov — APPROVED** [review]
🔧 : Just to be safe on all stages, I personally would also adjust the if-clauses in `#encodeHeaders()`, for example:

```
if (assignee != null && !assignee.isEmpty()) { ... }
```

(The same for `candidateGroups`.)

But it's up to you, take it or leave it 😉

**Comment by remcowesterhoud** [comment]
bors merge

**Comment by remcowesterhoud** [comment]
bors retry
